src/MELT/TestSink.cs

```diff
@@ -51,20 +64,75 @@ public void Write(WriteContext context)
             MessageLogged?.Invoke(context);
         }
 
-        public void BeginScope(BeginScopeContext context)
+        public IDisposable BeginScope(BeginScopeContext context)
         {
+            TestScope testScope;
+            lock (_scopeLock)
+            {
+                var oldScope = _currentScope;
+                testScope = new TestScope(this, oldScope);
+
+                //In principle, using Immutable collections here would prevent copying the collections
+                //But the number of scopes is low enough that it probably isn't worth the extra dependency
+                var newScopeData = oldScope.ScopeData.Prepend(new BeginScope(context)).ToList();
```

Comment:
I think if you use a insert it would potentially avoid the copy, not sure if the LINQ prepend has some sort of optimization.
But why not using a stack, like `ConcurrentStack` so you would also avoid to deal with the lock?
Or another alternative would be to maintain the link to the parent scope in the Scope itself, like implemented in the logging library: https://github.com/dotnet/runtime/blob/c7fda3b94a50297e6ece9db4ddc79ed039394dd3/src/libraries/Microsoft.Extensions.Logging/src/LoggerFactoryScopeProvider.cs#L66-L95 (we could potentially port this class over, it's small enough. In case, just include the original copyright notice).
Looking at that code also makes me wonder in which order we should return the scopes.

Reply:
Ah, I'm afraid I wasn't clear about this. The reason we have to either copy the collection or use an immutable collection here is that every log entry has a reference to the CurrentScope **at the time that log was created.** 
If we mutate the collection when we begin the scope, that will also change the scope of any previous log entries. 

We are already maintaining the link to parent scope in the `TestScope` class.

I'm agnostic about the scope return order. I did it this way so that it would behave like a stack. I think switching `Prepend` to `Append ` will reverse the order.

Reply:
Right, but if the scopes are in a linked list, can we not just store the reference to the last scope active at the time of capturing this log entry, instead of copying the list?